### PR TITLE
Fixes #10117

### LINKED ISSUE #10117: sizeof complains about illegal recursion on valid type definition
`sizeof` fails on the following forward list definition:
```nim
type
  SharedPtr[T] = object
    val: ptr tuple[atomicCounter: int, value: T]

  Node[T] = object
    value: T
    next: SharedPtr[Node[T]]

  ForwardList[T] = object
    first: SharedPtr[Node[T]]
    len: Natural
```

### PR DESCRIPTION
**Pull request by cooldome**
I have removed a check in `computeSizeAlign` as it rejects valid type definitions.
Fixes #10117.

### CODE REVIEW

**Comment by Araq** [comment]
Unfortunately that was required logic for ``tests/typerel/trectuples.nim``.

**Comment by cooldome** [comment]
At first I tried moving `illegal recursion` in tuple check into `semTuple`. It turned out to be both difficult and computationally demanding (requires a tree walk with cycle detector) and it needs to be done for every tuple defined.

So I decided to try opposite strategy - implement recursive tuples support in codegen. GC has no problems with recursive tuples already.

It turned out it is easier to support recursive tuples then detect and reject them in semantic pass

**Comment by Araq** [comment]
IIRC sigmatch also cannot handle recursive tuples.

**Comment by krux02** [comment]
I am against this new feature. First of all, because it is a new feature. It could introduce new bugs that we are not aware of. The issue it is supposed to fix, I cannot reproduce. And generally I don't see any real advantage to support this.

**Comment by cooldome** [comment]
@krux02: the PRcomes with a new test ` tests/destructor/trecursive.nim` that replicates the problem. Try to compile with latest Nim to get sizeof error.

**Comment by cooldome** [comment]
I will is recursive tuple to `semTuple`

**Comment by cooldome** [comment]
Alternative fix with cycle check is `semTuple` is now ready for review

File: compiler/types.nim [hunk 1]
```diff
@@ -1493,3 +1493,18 @@ proc typeMismatch*(conf: ConfigRef; info: TLineInfo, formal, actual: PType) =
...
     localError(conf, info, msg)
+
+proc isTupleRecursive*(t: PType, cycleDetector: var IntSet): bool =
+  if t == nil: return
+  case t.kind:
+    of tyTuple:
+      if cycleDetector.containsOrIncl(t.id):
+        return true
+      for i in  0..<t.len:
+        var copy = cycleDetector
```
Comment (Araq) [inline compiler/types.nim]: Why does it require a copy?

File: compiler/types.nim [hunk 2]
```diff
@@ -1493,3 +1493,18 @@ proc typeMismatch*(conf: ConfigRef; info: TLineInfo, formal, actual: PType) =
...
+        if isTupleRecursive(t[i], copy):
+          return true
+    of tyAlias, tyRef, tyPtr, tyGenericInst, tyVar, tyLent, tySink:
```
Comment (Araq) [inline compiler/types.nim]: Whitelisting seems dangerous, what if I try to create the recursion via ``tyArray`` or ``tySequence``?

Comment (LemonBoy) [reply compiler/types.nim]: It is dangerous, see #10565 for an example of undetected recursion via `tySequence`.

Comment (krux02) [reply compiler/types.nim]: I looked into it, and yes ``tyArray``, ``tyUncheckedArray`` and ``tySequence`` is missing. But it is not dangerous to use white listing here at all. If the logic is wrong (which is already the case with the current implementation) illegal recursive types will be allowed at this stage. But signature hashing will eventually crash the compiler. In my opinion that is just some form of a badly styled error message.

**Comment by Araq** [comment]
@krux02 took over.

**Pull request closed without merging**